File: exp-15/q2-kmp-string-matching/solution.js
const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

function computeLPS(p) {
    const m = p.length;
    const lps = new Array(m).fill(0);
    let len = 0, i = 1;
    while (i < m) {
        if (p[i] === p[len]) {
            len++;
            lps[i] = len;
            i++;
        } else if (len !== 0) {
            len = lps[len - 1];
        } else {
            lps[i] = 0;
            i++;
        }
    }
    return lps;
}

let lines = [];
rl.on('line', (line) => {
    lines.push(line);
    if (lines.length === 2) {
        const text = lines[0];
        const pattern = lines[1];
        const n = text.length, m = pattern.length;
        const lps = computeLPS(pattern);
        let i = 0, j = 0;
        let found = false;
        while (i < n) {
            if (pattern[j] === text[i]) {
                i++;
                j++;
            }
            if (j === m) {
                console.log(`Pattern found at index ${i - j}`);
                found = true;
                j = lps[j - 1];
            } else if (i < n && pattern[j] !== text[i]) {
                if (j !== 0) j = lps[j - 1];
                else i++;
            }
        }
        if (!found) console.log('not found');
        rl.close();
    }
});

File: exp-15/q2-kmp-string-matching/solution.py
def compute_lps(p):
    m = len(p)
    lps = [0] * m
    length, i = 0, 1
    while i < m:
        if p[i] == p[length]:
            length += 1
            lps[i] = length
            i += 1
        elif length != 0:
            length = lps[length - 1]
        else:
            lps[i] = 0
            i += 1
    return lps


text = input()
pattern = input()
n, m = len(text), len(pattern)
lps = compute_lps(pattern)

i = j = 0
found = False
while i < n:
    if pattern[j] == text[i]:
        i += 1
        j += 1
    if j == m:
        print(f"Pattern found at index {i - j}")
        found = True
        j = lps[j - 1]
    elif i < n and pattern[j] != text[i]:
        if j != 0:
            j = lps[j - 1]
        else:
            i += 1

if not found:
    print("not found")
